test/xbgpu/test_engine.py

```diff
@@ -236,116 +271,157 @@ def _create_heaps(
             but need not be contiguous.
         timestamp_step
             Timestamp step between each received heap processed.
-        n_ants, n_channels_per_substream, n_spectra_per_heap
+        n_ants, n_channels_per_substream, n_spectra_per_heap, frequency
             See :meth:`_create_heaps` for more info.
 
         Returns
         -------
         device_results
             Array of all GPU-generated data of shape
-            - (n_total_accumulations, n_channels_per_substream, n_baselines, COMPLEX)
+            - (len(corrprod_outputs), n_total_accumulations, n_channels_per_substream, n_baselines, COMPLEX)
+        n_accumulations_completed
+            Number of completed accumulations for each corrprod_output
         """
         max_packet_size = n_spectra_per_heap * N_POLS * COMPLEX * SAMPLE_BITWIDTH // 8 + PREAMBLE_SIZE
         max_heaps = n_ants * HEAPS_PER_FENGINE_PER_CHUNK * 10
-        feng_stream = self._make_feng(mock_recv_stream, max_packet_size, max_heaps)
+        feng_stream = self._make_feng(mock_recv_streams, max_packet_size, max_heaps)
+
+        acc_index_counts: list[dict[int, int]] = []
+        for corrprod_output in corrprod_outputs:
+            acc_index_counts.append(
+                dict.fromkeys(
+                    [batch_index // corrprod_output.heap_accumulation_threshold for batch_index in batch_indices], 0
+                )
+            )
+
+        # NOTE: The data generation generates an extra heap or several for the XPipeline
+        # with the smallest heap-accumulation-threshold - resulting in an incomplete final
+        # accumulation. Incomplete accumulations are purposefully dictated by `batch_indices`
+        # constructed by the caller.
+        for i, corrprod_output in enumerate(corrprod_outputs):
+            for batch_index in batch_indices:
+                adjusted_index = batch_index // corrprod_output.heap_accumulation_threshold
+                acc_index_counts[i][adjusted_index] += 1
+            # `adjusted_index` should be the last accumulation index for this corrprod
+            if acc_index_counts[i][adjusted_index] != corrprod_output.heap_accumulation_threshold:
+                acc_index_counts[i].pop(adjusted_index)
 
-        accumulation_indices_set = set()
         for batch_index in batch_indices:
             heaps = heap_factory(batch_index)
-            accumulation_indices_set.add(batch_index // heap_accumulation_threshold)
             await feng_stream.async_send_heaps(heaps, spead2.send.GroupMode.ROUND_ROBIN)
 
-        mock_recv_stream.stop()
-
-        # It is expected that the first packet will be a descriptor.
-        ig_recv = spead2.ItemGroup()
-        heap = await recv_stream.get()
-        items = ig_recv.update(heap)
-        assert len(list(items.values())) == 0, "This heap contains item values not just the expected descriptors."
+        for q in mock_recv_streams:
+            q.stop()
```

Comment:
It's more common in katgpucbf tests to use `queue` for a variable holding an InprocQueue.